### PR TITLE
Pcvl 548 svd evolve

### PR DESCRIPTION
**Pull request by melvinmathe**

### CODE REVIEW

File: perceval/simulators/simulator.py [hunk 1]
```diff
@@ -398,3 +396,17 @@ def evolve(self, input_state: Union[BasicState, StateVector]) -> StateVector:
 
         result_sv.normalize()
         return result_sv
+
+    def evolve_svd(self, svd: Union[SVDistribution, StateVector, BasicState]) -> SVDistribution:
+        """Compute the SVDistribution evolved through a Linear Optical circuit"""
```
Comment (ericbrts) [inline perceval/simulators/simulator.py:401]: Could you write a complete docstring for this method (you can check other methods of this class)? Actually, the code reference of this class is autogenerated as user documentation.

File: perceval/simulators/simulator.py [hunk 2]
```diff
@@ -398,3 +396,17 @@ def evolve(self, input_state: Union[BasicState, StateVector]) -> StateVector:
 
         result_sv.normalize()
         return result_sv
+
+    def evolve_svd(self, svd: Union[SVDistribution, StateVector, BasicState]) -> SVDistribution:
```
Comment (ericbrts) [inline perceval/simulators/simulator.py:400]: Could you take into account an optional callback (like in probs_svd)?

File: perceval/simulators/simulator.py [hunk 3]
```diff
@@ -398,3 +396,17 @@ def evolve(self, input_state: Union[BasicState, StateVector]) -> StateVector:
...
+            return SVDistribution(self.evolve(svd))
+
+        # If it's actually an SVD
+
+        new_svd = SVDistribution()
+        for sv, p in svd.items():
+            new_sv = self.evolve(sv)
+            new_svd[new_sv] = p
+
+        return new_svd
```
Comment (ericbrts) [inline perceval/simulators/simulator.py:412]: In order to be consistent with `probs_svd` call, this one should support post-selection too.
(We just discussed this orally, sorry, it's not in the Jira ticket, sorry)

**Review by ericbrts — CHANGES_REQUESTED** [review]

File: perceval/simulators/simulator.py [hunk 4]
```diff
@@ -397,4 +409,32 @@ def evolve(self, input_state: Union[BasicState, StateVector]) -> StateVector:
...
+        if not isinstance(svd, SVDistribution):
+            return SVDistribution(self.evolve(svd))
+
+        # If it's actually an SVD
+
+        new_svd = SVDistribution()
+        for idx, (sv, p) in enumerate(svd.items()):
+            if min(sv.n) >= self._min_detected_photons:
+                new_sv = self._post_select_on_statevector(self.evolve(sv))
+                new_svd[new_sv] = p
```
Comment (ericbrts) [inline perceval/simulators/simulator.py:432]: I think
`new_svd[new_sv] += p` would fix corner cases where `new_sv` could be the result of several post_selected evolutions.

File: tests/test_simulator.py [hunk 5]
```diff
@@ -282,3 +282,13 @@ def test_evovle_phase():
     simu.set_circuit(c)
     output_sv = simu.evolve(input_state)
     assert output_sv[BasicState([1, 1])] == pytest.approx(complex(math.sqrt(2)/4, math.sqrt(6)/4))
+
+
+def test_simulator_evolve_svd():
```
Comment (ericbrts) [inline tests/test_simulator.py:287]: Do you think you can add a more complex test for `evolve_svd` with post-selection?

**Review by ericbrts — APPROVED** [review]
